### PR TITLE
bpo-40217: Clean code in PyType_FromSpec_Alloc and add NEWS entry

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue40217

### CODE REVIEW

**Comment by pablogsal** [comment]
Some minor cleanups were missing before #19414 was merged

File: Objects/typeobject.c [hunk 1]
```diff
@@ -1031,25 +1031,29 @@ PyType_FromSpec_Alloc(PyTypeObject *type, Py_ssize_t nitems)
...
+    }
+    else {
         obj = (PyObject *)PyObject_MALLOC(size);
+    }
 
-    if (obj == NULL)
+    if (obj == NULL) {
         return PyErr_NoMemory();
-
-    obj = obj;
```
Comment (vstinner) [inline Objects/typeobject.c:1042]: Would you mind to keep this instruction? I love it :-D (I'm kidding you, please remove it!)

File: Objects/typeobject.c [hunk 2]
```diff
@@ -3066,7 +3070,11 @@ PyType_FromSpecWithBases(PyType_Spec *spec, PyObject *bases)
             *
             * We store the user-provided traverse function at the end of the type
             * (we have allocated space for it) so we can call it from our
-            * PyType_FromSpec_tp_traverse wrapper. */
+            * PyType_FromSpec_tp_traverse wrapper.
+            *
+            * Check bpo-40217 for more information and rationale about this issue.
```
Comment (vstinner) [inline Objects/typeobject.c:3075]: Your phrasing is good and nice. But my suggestion was just to add "bpo-40217: " at the beginning of the comment. It's up to you.

File: Misc/NEWS.d/next/C API/2020-04-27-14-00-38.bpo-40217.sgn6c8.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Ensure that instances of types created with
+:c:func:`PyType_FromSpecWithBases` will visit its class object when
+traversing references in the garbage collector (implemented as an extension
+of the provided :c:member:`~PyTypeObject.tp_traverse`). Patch by Pablo
+Galindo
```
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2020-04-27-14-00-38.bpo-40217.sgn6c8.rst]:
```suggestion
Galindo.
```

File: Objects/typeobject.c (hunk 2 above)
Comment (pablogsal) [reply Objects/typeobject.c:3075]: I personally like this way a bit more, bit I can move it to the top of you want

File: Objects/typeobject.c (hunk 1 above)
Comment (pablogsal) [reply Objects/typeobject.c:1042]: Maybe we should add some `volatile` 😉

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**